src/System.Numerics.Tensors/System/Numerics/Tensor.cs

```diff
@@ -0,0 +1,1197 @@
+            {
+                if (left.Dimensions[i] != result.Dimensions[i] || right.Dimensions[i] != result.Dimensions[i])
+                {
+                    throw new ArgumentException("Operands and result must have matching dimensions");
+                }
+            }
+        }
+
+        internal static void ValidateBinaryArgs<T>(Tensor<T> left, Tensor<T> right, Tensor<bool> result)
+        {
+            if (left == null)
+            {
+                throw new ArgumentNullException(nameof(left));
+            }
+
+            if (right == null)
+            {
+                throw new ArgumentNullException(nameof(right));
+            }
+
+            if (result == null)
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+
+            // TODO: Support "compatible sizes" AKA "broadcasting"
+
+            if (left.Rank != result.Rank || right.Rank != result.Rank || left.Length != result.Length || right.Length != result.Length)
+            {
+                throw new ArgumentException("Operands and result must have matching dimensions");
+            }
+
+            for (int i = 0; i < result.Rank; i++)
+            {
+                if (left.Dimensions[i] != result.Dimensions[i] || right.Dimensions[i] != result.Dimensions[i])
+                {
+                    throw new ArgumentException("Operands and result must have matching dimensions");
+                }
+            }
+        }
+
+        internal static void ValidateArgs<T>(Tensor<T> tensor)
+        {
+            if (tensor == null)
+            {
+                throw new ArgumentNullException(nameof(tensor));
+            }
+        }
+
+        internal static void ValidateArgs<T>(Tensor<T> tensor, Tensor<T> result)
+        {
+            if (tensor == null)
+            {
+                throw new ArgumentNullException(nameof(tensor));
+            }
+
+            if (result == null)
+            {
+                throw new ArgumentNullException(nameof(result));
+            }
+
+            if (tensor.Rank != result.Rank || tensor.Length != result.Length)
+            {
+                throw new ArgumentException("Operands and result must have matching dimensions");
+            }
+
+            for (int i = 0; i < result.Rank; i++)
+            {
+                if (tensor.Dimensions[i] != result.Dimensions[i])
+                {
+                    throw new ArgumentException("Operands and result must have matching dimensions");
+                }
+            }
+        }
+        
+    }
+
+    // When we cross-compile for frameworks that expose ICloneable this must implement ICloneable as well.
+    public class Tensor<T>: IList, ICollection, IEnumerable, IStructuralComparable, IStructuralEquatable
+    {
+        private readonly T[] backingArray;
+        private readonly int[] dimensions;
+        private IReadOnlyList<int> readOnlyDimensions;
```

Comment:
If everything's readonly and it's just these few fields, could `Tensor<T>` be an immutable struct instead of a class?